core/src/Cuda/Kokkos_Cuda_UniqueToken.hpp

```diff
@@ -83,46 +83,79 @@ class UniqueToken<Cuda, UniqueTokenScope::Global> {
 
   /// \brief upper bound for acquired values, i.e. 0 <= value < size()
   KOKKOS_INLINE_FUNCTION
-  size_type size() const noexcept { return m_count; }
+  size_type size() const noexcept { return m_locks.extent(0); }
 
-  /// \brief acquire value such that 0 <= value < size()
+ private:
   KOKKOS_INLINE_FUNCTION
-  size_type acquire() const {
-    const Kokkos::pair<int, int> result =
-        Kokkos::Impl::concurrent_bitset::acquire_bounded(
-            m_buffer, m_count, Kokkos::Impl::clock_tic() % m_count);
-
-    if (result.first < 0) {
-      Kokkos::abort(
-          "UniqueToken<Cuda> failure to acquire tokens, no tokens available");
+  size_type impl_acquire() const {
+    int idx = blockIdx.x * (blockDim.x * blockDim.y) +
+              threadIdx.y * blockDim.x + threadIdx.x;
+    idx = idx % size();
+#if defined(KOKKOS_ARCH_KEPLER) || defined(KOKKOS_ARCH_PASCAL) || \
+    defined(KOKKOS_ARCH_MAXWELL)
+    unsigned int mask        = __activemask();
+    unsigned int active      = __ballot_sync(mask, 1);
+    unsigned int done_active = 0;
+    bool done                = false;
+    while (active != done_active) {
+      if (!done) {
+        if (Kokkos::atomic_compare_exchange(&m_locks(idx), 0, 1) == 0) {
+          done = true;
+        } else {
+          idx += blockDim.y * blockDim.x + 1;
+          idx = idx % size();
+        }
+      }
+      done_active = __ballot_sync(mask, done ? 1 : 0);
+    }
+#else
+    while (Kokkos::atomic_compare_exchange(&m_locks(idx), 0, 1) == 1) {
+      idx += blockDim.y * blockDim.x + 1;
+      idx = idx % size();
     }
+#endif
+// Make sure that all writes in the previous lock owner are visible to me
+#ifdef KOKKOS_ENABLE_IMPL_DESUL_ATOMICS
+    desul::atomic_thread_fence(desul::MemoryOrderAcquire(),
+                               desul::MemoryScopeDevice());
```

Comment:
What writes?  To `&m_locks(idx)`?  If so is it really needed.  It looks like all accesses to `m_lock` are atomic operations.

Reply:
writes happening between acquire and release of the token ??